cpp/src/arrow/util/checked_cast.h

```diff
@@ -0,0 +1,42 @@
+// Licensed to the Apache Software Foundation (ASF) under one
+// or more contributor license agreements.  See the NOTICE file
+// distributed with this work for additional information
+// regarding copyright ownership.  The ASF licenses this file
+// to you under the Apache License, Version 2.0 (the
+// "License"); you may not use this file except in compliance
+// with the License.  You may obtain a copy of the License at
+//
+//   http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+#ifndef ARROW_CAST_H
+#define ARROW_CAST_H
+
+#include <type_traits>
+
+namespace arrow {
+
+template <typename OutputType, typename InputType>
+OutputType checked_cast(InputType&& value) {
+  static_assert(std::is_class<typename std::remove_pointer<
+                    typename std::remove_reference<InputType>::type>::type>::value,
+                "checked_cast input type must be a class");
+  static_assert(std::is_class<typename std::remove_pointer<
+                    typename std::remove_reference<OutputType>::type>::type>::value,
+                "checked_cast output type must be a class");
+#ifdef NDEBUG
+  return static_cast<OutputType>(value);
+#else
+  return dynamic_cast<OutputType>(value);
```

Comment:
This header leaks in the public API, so if downstream users do not define NDEBUG then they will have dynamic_casts wherever this is used

Reply:
Hm, that's kind of the point since that would mean that someone is building a debug version of arrow. Are we trying to allow for downstream developers that do not define `NDEBUG` and expect to get non-debug builds for the dependencies they are building themselves?

Reply:
Indeed if people are building in debug mode they probably want the extra safety checks.